### PR TITLE
Approximates chatmessage's height and width value

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

This PR removes call to MeasureText(), effectively making creating chatmessages synchronous.
This fixes a few possible runtimes and weird undefined behaviors.

~~`CHAR_WIDTH` is (`width` / 1000) * `font_size`~~
~~Where width is width as calculated by FontForge. `width` / 1000 is the conversion between FontForge's em-size value and 1000 em-size is just 1 em.~~
~~Check the FontForge's docs for better explanation. https://fontforge.org/docs/~~
~~`APPROX_HEIGHT` is `font_size` * 1.7~~
~~Where 1.7 is approximation, based on comparing results of the MeasureText() and used font.~~

~~The values from FontForge are saved in static list. If something isn't on the list we assume the worst and give it width of 1000.~~

The values from font forge were inacurate. The approx width compared to measured one differed by about 20%.
The more naive solution has been implemented - measure cache of some sort. Every client should have the same viewport so the measurements between clients should be the same. Thus, once, we calculate values for every character and keep using them.
After testing it is about ~1% inacurate (which is huge improvement over 20%).

~~Additionally the values are cached now per client plus every uncached character will be calculated for all the clients.~~

I changed my mind. I precalculated values for first 1500 characters for unicode. That should cover all the characters you would ever need. Anything that isn't included gets replaced with "?". 
If you ever need a new character you can modify json with the value.

## Why It's Good For The Game

Less chatmessage bugs and better performance.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

obraz

obraz
obraz

obraz

obraz

obraz

obraz

NEW CONTENT BELLOW
(old is still relevant though)
image
sunglasses emoji was replaced with "?"

</details>

## Changelog
:cl:
refactor: refactored how chatmessage width and height is calculated
/:cl:

### CODE REVIEW

File: code/datums/chatmessage.dm [hunk 1]
```diff
@@ -561,6 +658,14 @@ GLOBAL_LIST_INIT(job_colors_pastel, list(
...
 	fadertimer = addtimer(CALLBACK(src, .proc/end_of_life), duration, TIMER_STOPPABLE|TIMER_DELETE_ME, SSrunechat)
 
+/datum/chatmessage/proc/approx_str_width(var/string, var/font_size = DEFAULT_FONT_SIZE)
+	var/value = 0
+	for(var/i in 1 to length(string))
+		var/size = letters[string[i]]
+		if(!size)
+			size = 1000
+		value += CHAR_WIDTH(size, font_size)
+	return value
```
Comment (PowerfulBacon) [inline code/datums/chatmessage.dm]: Is having a for loop on every character actually an optimisation over what its doing now?

Comment (Archanial) [reply code/datums/chatmessage.dm]: This is very simple linear loop with access to the assoc list.
The overall cost is O(n), I doubt it's any worse.

Comment (ike709) [reply code/datums/chatmessage.dm]: Faster than MeasureText() for sure.

**Comment by Archanial** [comment]
Issue with this is that it's not very accurate to display pixels without knowing the DPI of someone's screen.
When assuming the DPI of 75 (easiest to assume) we're losing 1 px on average on ~96 DPI screens. 
This needs a way to calculate either per client dpi multiplier or actual dpi.

**Comment by ike709** [comment]
Could simply give people a game setting to set their DPI, but there's probably a lot of players who don't know what DPI is.

**Comment by Archanial** [comment]
I have 2 ideas for that,
you could either use TGUI to check someone's dpi
or do some MeasureText for clients on login

**Comment by LemonInTheDark** [comment]
> Every client should have the same viewport

what does this mean?

**Comment by ike709** [comment]
Wasn't this more or less done?

**Comment by Archanial** [comment]
we so back

**Comment by LemonInTheDark** [comment]
what?

**Comment by Archanial** [comment]
I can't reopen so #11947 is new pr for this

**Pull request closed without merging**